server/auth.py

```diff
@@ -1,67 +1,39 @@
-
-
-@app.route('/auth', methods=['GET'])
 def auth():
-    authInfo = None
-    for key, val in request.cookies.items():
-        if '_shibsession_' in key:
-            authInfo = key + val
-            break
-    if authInfo:
-        # Convert to bytes for Python 3, assume always ASCII
-        auth_info = bytearray(authInfo, 'ascii')
-        # Salt is not necessary since we are hashing a session token
-        hmac_token = hashlib.pbkdf2_hmac('sha256', auth_info, b'salt', 100000)
-        authToken = binascii.hexlify(hmac_token)
-        db.set('authToken:%s' % authToken, 1)
-        return authToken
-    else:
-        return jsonify({'error': 'no shibboleth cookie'}), 400
-
-
-@app.route('/validate/<string:token>', methods=['GET'])
-def validate(token):
-    debug = app.debug or app.testing
-    if not request.base_url.startswith('https') and not debug:
-        return jsonify({'status': 'insecure access over http'}), 401
-    if db.exists('authToken:%s' % token):
-        return jsonify({'status': 'valid'})
-    else:
-        return jsonify({'status': 'invalid'}), 401
-
-
-def auth_decorator(f):
-    @wraps(f)
-    def wrapper(*args, **kwds):
-        if db.exists('authToken:%s' % request.args['authToken']):
-            return f(*args, **kwds)
-        else:
-            return jsonify({'error': 'invalid auth token'}), 401
-
-    return wrapper
+    def _auth(f):
+        @wraps(f)
+        def __auth(*args, **kwargs):
+            authorization = request.headers.get('Authorization')
+            if authorization and ' ' in authorization:
+                auth_type, token = authorization.split()
+                if auth_type == 'Bearer':  # Only validate if Authorization header type is Bearer
+                    try:
+                        body = {'token': token}
+                        headers = {'Authorization': 'Bearer {}'.format(token)}
+                        data = requests.post(
+                            url='https://platform.pennlabs.org/accounts/introspect/',
+                            headers=headers,
+                            data=body
+                        )
+                        if data.status_code == 200:  # Access token is valid
+                            data = data.json()
+                            account = Account.query.filter_by(pennid=data['user']['pennid']).first()
+                            return f(account)
+                        else:
+                            return jsonify({'error': 'Invalid token'}), 401
+                    except requests.exceptions.RequestException:  # Can't connect to platform
+                        # Throw a 403 because we can't verify the incoming access token so we
+                        # treat it as invalid. Ideally platform will never go down, so this
+                        # should never happen.
+                        return jsonify({'error': 'Unable to connect to Platform'}), 401
+            else:
```

Comment:
In the long-term, the only route I can think of like that is `homepage`, since we want to be able to display a homepage to users that aren't logged in. I've added a `nullable` field to the decorator, which if true, will call the function without it an account if no account is found or if a token is not passed or is invalid. Works as follows:
```
@app.route('/housing', methods=['GET'])
@auth(nullable=True)
def get_housing_info():
    ... Continue, g.account may be None
```